WalletWasabi/Stores/IoManager.cs

```diff
@@ -0,0 +1,374 @@
+using System.IO;
+using System.Text;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using Nito.AsyncEx;
+
+namespace WalletWasabi.Stores
+{
+	/// <summary>
+	/// Safely manager file operations.
+	/// </summary>
+	public class IoManager
+	{
+		public string OriginalFilePath { get; }
+		public string OldFilePath { get; }
+		public string NewFilePath { get; }
+		public string DigestFilePath { get; }
+
+		/// <summary>
+		/// Use the random index of the line to create digest faster. -1 is special value, it means the last character. If null then hash whole file.
+		/// </summary>
+		private int? DigestRandomIndex { get; }
+
+		public string FileName { get; }
+		public string FileNameWithoutExtension { get; }
+		public AsyncMutex Mutex { get; }
+
+		private const string OldExtension = ".old";
+		private const string NewExtension = ".new";
+		private const string DigestExtension = ".dig";
+
+		/// <param name="digestRandomIndex">Use the random index of the line to create digest faster. -1 is special value, it means the last character. If null then hash whole file.</param>
+		public IoManager(string filePath, int? digestRandomIndex = null)
+		{
+			DigestRandomIndex = digestRandomIndex;
+			OriginalFilePath = Guard.NotNullOrEmptyOrWhitespace(nameof(filePath), filePath, trim: true);
+			OldFilePath = $"{OriginalFilePath}{OldExtension}";
+			NewFilePath = $"{OriginalFilePath}{NewExtension}";
+			DigestFilePath = $"{OriginalFilePath}{DigestExtension}";
+
+			FileName = Path.GetFileName(OriginalFilePath);
+			var shortHash = HashHelpers.GenerateSha256Hash(OriginalFilePath).Substring(0, 7);
+			FileNameWithoutExtension = Path.GetFileNameWithoutExtension(OriginalFilePath);
+
+			// https://docs.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=netframework-4.8
+			// On a server that is running Terminal Services, a named system mutex can have two levels of visibility.
+			// If its name begins with the prefix "Global\", the mutex is visible in all terminal server sessions.
+			// If its name begins with the prefix "Local\", the mutex is visible only in the terminal server session where it was created.
+			// In that case, a separate mutex with the same name can exist in each of the other terminal server sessions on the server.
+			// If you do not specify a prefix when you create a named mutex, it takes the prefix "Local\".
+			// Within a terminal server session, two mutexes whose names differ only by their prefixes are separate mutexes,
+			// and both are visible to all processes in the terminal server session.
+			// That is, the prefix names "Global\" and "Local\" describe the scope of the mutex name relative to terminal server sessions, not relative to processes.
+			Mutex = new AsyncMutex($"{FileNameWithoutExtension}-{shortHash}");
+		}
+
+		#region IoOperations
+
+		public void DeleteMe()
+		{
+			if (File.Exists(OriginalFilePath))
```

Comment:
This can fail in rare situations when the file exists when we ask but doesn't exist anymore when we request the deletion (the user could move the files or delete it). In case the OS locks the file this could raise an IOException. I think instead of doing the Tester-Doer we could try to delete it into a try-catch block.